File: ldk/csharp/OliveHelpsLDK/OliveHelpsLDK/LoopSensorFacade.cs
using System;
using System.Threading.Tasks;
using Grpc.Core;
using Grpc.Core.Interceptors;
using OliveHelpsLDK.Browser;
using OliveHelpsLDK.Clipboard;
using OliveHelpsLDK.Cursor;
using OliveHelpsLDK.Filesystem;
using OliveHelpsLDK.Hover;
using OliveHelpsLDK.Keyboard;
using OliveHelpsLDK.Logging;
using OliveHelpsLDK.Network;
using OliveHelpsLDK.Process;
using OliveHelpsLDK.Storage;
using OliveHelpsLDK.UI;
using OliveHelpsLDK.Whispers;
using OliveHelpsLDK.Window;

namespace OliveHelpsLDK
{
    internal class LoopSensorFacade : ILoopSensors
    {
        internal async Task Connect(ConnectionInfo connectionInfo, Session session, ILogger logger)
        {
            var address = connectionInfo.Network == "unix"
                ? $"unix://{connectionInfo.Address}"
                : connectionInfo.Address;

            var channel = new Channel(address, ChannelCredentials.Insecure);
            await channel.ConnectAsync();
            var interceptedChannel = channel.Intercept(new ExceptionLoggingInterceptor(logger));

            Console.Error.WriteLine("[DEBUG] GRPC Channel Connected");
            Whisper = new WhisperClient(interceptedChannel, session, logger);
            Clipboard = new ClipboardClient(interceptedChannel, session, logger);
            Filesystem = new FilesystemClient(interceptedChannel, session, logger);
            Cursor = new CursorClient(interceptedChannel, session, logger);
            Keyboard = new KeyboardClient(interceptedChannel, session, logger);
            Network = new NetworkClient(interceptedChannel, session, logger);
            Process = new ProcessClient(interceptedChannel, session, logger);
            Browser = new BrowserClient(interceptedChannel, session, logger);
            Hover = new HoverClient(interceptedChannel, session, logger);
            Window = new WindowClient(interceptedChannel, session, logger);
            Storage = new StorageClient(interceptedChannel, session, logger);
            UI = new UIClient(interceptedChannel, session, logger);
        }

        public IWhisperSensor Whisper { get; private set; }

        public IClipboardSensor Clipboard { get; private set; }

        public IFilesystemSensor Filesystem { get; private set; }

        public ICursorSensor Cursor { get; private set; }

        public IKeyboardSensor Keyboard { get; private set; }

        public INetworkSensor Network { get; private set; }

        public IProcessSensor Process { get; private set; }

        public IStorageSensor Storage { get; private set; }

        public IBrowserSensor Browser { get; private set; }

        public IHoverSensor Hover { get; private set; }

        public IWindowSensor Window { get; private set; }

        public IUISensor UI { get; private set; }
    }
}

File: ldk/node/src/loopSensorFacade.ts
import { ConnInfo } from './grpc/broker_pb';
import { Session } from './grpc/session_pb';
import { LoopSensors } from './loopSensors';
import WhisperClient from './loopClients/whisperClient';
import StorageClient from './loopClients/storageClient';
import KeyboardClient from './loopClients/keyboardClient';
import { ClipboardClient } from './loopClients/clipboardClient';
import { CursorClient } from './loopClients/cursorClient';
// import { HoverClient } from './loopClients/hoverClient';
import { FileSystemClient } from './loopClients/fileSystemClient';
import { ProcessClient } from './loopClients/processClient';
// import { WindowClient } from './loopClients/windowClient';
// import { BrowserClient } from './loopClients/browserClient';
import { NetworkClient } from './loopClients/networkClient';
import { UIClient } from './loopClients/uiClient';
import { Logger } from './logging';

/**
 * @internal
 */
export default class LoopSensorFacade implements LoopSensors {
  private logger: Logger;

  public whisper: WhisperClient = new WhisperClient();

  public storage: StorageClient = new StorageClient();

  public keyboard: KeyboardClient = new KeyboardClient();

  public clipboard: ClipboardClient = new ClipboardClient();

  public cursor: CursorClient = new CursorClient();

  public fileSystem: FileSystemClient = new FileSystemClient();

  public process: ProcessClient = new ProcessClient();

  public ui: UIClient = new UIClient();

  // These sensors are not yet implemented.
  // public hover: HoverClient = new HoverClient();

  // public window: WindowClient = new WindowClient();

  // public browser: BrowserClient = new BrowserClient();

  constructor(logger: Logger) {
    this.logger = logger;
  }

  public network: NetworkClient = new NetworkClient();

  public connect(
    connInfo: ConnInfo.AsObject,
    session: Session.AsObject,
  ): Promise<void[]> {
    return Promise.all([
      // These sensors are not yet implemented.
      // this.browser.connect(connInfo, session, this.logger),
      // this.hover.connect(connInfo, session, this.logger),
      // this.window.connect(connInfo, session, this.logger),
      this.whisper.connect(connInfo, session, this.logger),
      this.storage.connect(connInfo, session, this.logger),
      this.clipboard.connect(connInfo, session, this.logger),
      this.keyboard.connect(connInfo, session, this.logger),
      this.process.connect(connInfo, session, this.logger),
      this.cursor.connect(connInfo, session, this.logger),
      this.fileSystem.connect(connInfo, session, this.logger),
      this.network.connect(connInfo, session, this.logger),
    ]);
  }
}
